File: sdk/python/src/burrow/_transport.py
"""gRPC transport. Internal: the public surface in `sandbox.py` is hand-written
and does not mirror the RPC shapes."""

from __future__ import annotations

import os
import re
from typing import Any, Iterator, Optional

import grpc

from ._pb import api_pb2_grpc
from .errors import BurrowError

# Sentinel distinguishing "use the transport default" from "no deadline at
# all", which long-lived streams like watches need.
_DEFAULT = object()

# The connection options every public entry point accepts alongside its own.
_TRANSPORT_KEYS = ("endpoint", "api_key", "tls", "timeout")


def transport_options(options: dict) -> dict:
    """The transport's share of a mixed option bag."""
    return {key: options[key] for key in _TRANSPORT_KEYS if key in options}


def _is_loopback_host(host: str) -> bool:
    """Whether `host` is plainly this machine, not some other one on the network."""
    bare = host.lstrip("[").rstrip("]")
    return (
        bare == "localhost"
        or bare == "::1"
        or bool(re.fullmatch(r"127\.\d{1,3}\.\d{1,3}\.\d{1,3}", bare))
    )


class Transport:
    def __init__(
        self,
        endpoint: Optional[str] = None,
        api_key: Optional[str] = None,
        tls: Optional[bool] = None,
        timeout: Optional[float] = 60.0,
    ):
        raw = endpoint or os.environ.get("BURROW_ENDPOINT") or "localhost:7070"
        secure = tls if tls is not None else raw.startswith("https://")
        address = re.sub(r"^https?://", "", raw)

        # Host this client dials, without its port. A published port whose node
        # advertises no address of its own is reachable here, the single-node
        # case.
        self.host = re.sub(r":\d+$", "", address) or "localhost"
        self.timeout = timeout
        self._api_key = api_key or os.environ.get("BURROW_API_KEY")
        # Handles sharing this connection; the channel outlives the first
        # release.
        self._refs = 1

        # A bare `host:port` endpoint leaves `tls` off, which is fine with no
        # credential in play but an easy way to ship an API key in cleartext to
        # whatever `BURROW_ENDPOINT` is set to in production. Refused rather
        # than warned, unless the destination is obviously the caller's own
        # machine or `tls`/the endpoint scheme said so explicitly.
        if (
            self._api_key
            and not secure
            and tls is None
            and not raw.startswith("http://")
            and not _is_loopback_host(self.host)
        ):
            raise BurrowError(
                f"refusing to send an API key over a plaintext connection to "
                f"{self.host}; pass tls explicitly if that's intended",
                "invalid_argument",
            )

        options = [
            # Exec and file transfers can carry large payloads; the 4MB default
            # turns a big file into a confusing RESOURCE_EXHAUSTED.
            ("grpc.max_receive_message_length", 64 * 1024 * 1024),
            ("grpc.max_send_message_length", 64 * 1024 * 1024),
        ]
        if secure:
            self._channel = grpc.secure_channel(
                address, grpc.ssl_channel_credentials(), options
            )
        else:
            self._channel = grpc.insecure_channel(address, options)
        self._stub = api_pb2_grpc.BurrowStub(self._channel)

    def _metadata(self):
        if self._api_key:
            return (("authorization", f"Bearer {self._api_key}"),)
        return ()

    def _timeout(self, timeout: Any) -> Optional[float]:
        return self.timeout if timeout is _DEFAULT else timeout

    def unary(self, method: str, request: Any, timeout: Any = _DEFAULT) -> Any:
        try:
            return getattr(self._stub, method)(
                request, timeout=self._timeout(timeout), metadata=self._metadata()
            )
        except grpc.RpcError as err:
            raise BurrowError.from_grpc(err) from None

    def server_stream(
        self, method: str, request: Any, timeout: Any = _DEFAULT
    ) -> Iterator[Any]:
        return _iterate(self.open_server_stream(method, request, timeout))

    def open_server_stream(
        self, method: str, request: Any, timeout: Any = _DEFAULT
    ) -> Any:
        """Server-streaming call, returned as the raw call: an iterator of
        responses that also cancels (`call.cancel()`), which stopping a watch
        from another thread needs."""
        return getattr(self._stub, method)(
            request, timeout=self._timeout(timeout), metadata=self._metadata()
        )

    def client_stream(
        self, method: str, requests: Iterator[Any], timeout: Any = _DEFAULT
    ) -> Any:
        """Client-streaming call whose single response is awaited (upload)."""
        try:
            return getattr(self._stub, method)(
                requests, timeout=self._timeout(timeout), metadata=self._metadata()
            )
        except grpc.RpcError as err:
            raise BurrowError.from_grpc(err) from None

    def duplex(
        self, method: str, requests: Iterator[Any], timeout: Any = _DEFAULT
    ) -> Any:
        """Bidirectional call. Returns the raw call: an iterator of responses
        that also cancels (`call.cancel()`), which a terminal needs."""
        return getattr(self._stub, method)(
            requests, timeout=self._timeout(timeout), metadata=self._metadata()
        )

    def retain(self) -> "Transport":
        """Claims a share of this connection. Undone by one `close`."""
        self._refs += 1
        return self

    def close(self) -> None:
        """Releases one share; the channel closes when the last one goes."""
        self._refs -= 1
        if self._refs <= 0:
            self._channel.close()


def _iterate(call: Any) -> Iterator[Any]:
    """Adapts a streaming call so its errors raise as `BurrowError`, and an
    abandoned loop (break, exception) cancels the call rather than leaking it."""
    try:
        for msg in call:
            yield msg
    except grpc.RpcError as err:
        raise BurrowError.from_grpc(err) from None
    finally:
        call.cancel()

File: sdk/typescript/src/transport.ts
/**
 * gRPC transport. Internal: the public surface in `sandbox.ts` is hand-written
 * and does not mirror the RPC shapes.
 *
 * Protos are loaded at runtime with `@grpc/proto-loader` rather than compiled
 * ahead of time, so the package needs no code generation step and ships the
 * same `.proto` files the daemons are built from.
 */

import { createRequire } from "node:module";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import * as grpc from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";

import { BurrowError } from "./errors.js";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));

/** Resolves `proto/` whether running from `src/` or from built `dist/`. */
function protoDir(): string {
  for (const candidate of [
    resolve(here, "../proto"),
    resolve(here, "../../proto"),
  ]) {
    try {
      require.resolve(resolve(candidate, "api.proto"));
      return candidate;
    } catch {
      // Not this one; try the next.
    }
  }
  // require.resolve rejects non-JS files in some setups, so fall back to the
  // layout of a published package.
  return resolve(here, "../proto");
}

export interface TransportOptions {
  /** Orchestrator address, e.g. `localhost:7070` or `http://localhost:7070`. */
  endpoint?: string;
  /** Bearer token, sent as `authorization: Bearer <token>` on every call. */
  apiKey?: string;
  /** Use TLS. Defaults to true for `https://` endpoints, false otherwise. */
  tls?: boolean;
  /** Default per-call deadline in milliseconds. */
  timeoutMs?: number;
}

type AnyClient = grpc.Client & Record<string, Function>;

/** Whether `host` is plainly this machine, not some other one on the network. */
function isLoopbackHost(host: string): boolean {
  const bare = host.replace(/^\[/, "").replace(/\]$/, "");
  return (
    bare === "localhost" ||
    bare === "::1" ||
    /^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(bare)
  );
}

/**
 * Cancels `call` when `signal` aborts, and returns a cleanup function.
 *
 * grpc-js has no `AbortSignal`, but a cancelled call surfaces as `CANCELLED`,
 * which is what an aborted request should look like.
 */
function bindSignal(call: any, signal: AbortSignal | undefined): () => void {
  if (!signal) return () => {};
  if (signal.aborted) {
    call.cancel();
    return () => {};
  }
  const onAbort = () => call.cancel();
  signal.addEventListener("abort", onAbort, { once: true });
  return () => signal.removeEventListener("abort", onAbort);
}

export class Transport {
  private readonly client: AnyClient;
  readonly timeoutMs: number;
  /**
   * Host this client dials, without its port. A published port whose node
   * advertises no address of its own is reachable here, the single-node case.
   */
  readonly host: string;
  private readonly apiKey?: string;
  /** Handles sharing this connection; the channel outlives the first release. */
  private refs = 1;

  constructor(options: TransportOptions = {}) {
    const raw =
      options.endpoint ?? process.env.BURROW_ENDPOINT ?? "localhost:7070";
    const secure = options.tls ?? raw.startsWith("https://");
    const address = raw.replace(/^https?:\/\//, "");

    this.host = address.replace(/:\d+$/, "") || "localhost";
    this.timeoutMs = options.timeoutMs ?? 60_000;
    this.apiKey = options.apiKey ?? process.env.BURROW_API_KEY;

    // A bare `host:port` endpoint leaves `tls` false, which is fine with no
    // credential in play but an easy way to ship an API key in cleartext to
    // whatever `BURROW_ENDPOINT` is set to in production. Refused rather
    // than warned, unless the destination is obviously the caller's own
    // machine or `tls`/the endpoint scheme said so explicitly.
    if (
      this.apiKey &&
      !secure &&
      options.tls === undefined &&
      !raw.startsWith("http://") &&
      !isLoopbackHost(this.host)
    ) {
      throw new BurrowError(
        `refusing to send an API key over a plaintext connection to ${this.host}; ` +
          `pass tls explicitly if that's intended`,
        "invalid_argument",
      );
    }

    const definition = protoLoader.loadSync("api.proto", {
      includeDirs: [protoDir()],
      keepCase: false,
      longs: Number,
      enums: String,
      defaults: true,
      oneofs: true,
    });
    const pkg = grpc.loadPackageDefinition(definition) as any;
    const Ctor = pkg.burrow.api.v1.Burrow;

    this.client = new Ctor(
      address,
      secure
        ? grpc.credentials.createSsl()
        : grpc.credentials.createInsecure(),
      {
        // Exec and file transfers can carry large payloads; the 4MB default
        // turns a big file into a confusing RESOURCE_EXHAUSTED.
        "grpc.max_receive_message_length": 64 * 1024 * 1024,
        "grpc.max_send_message_length": 64 * 1024 * 1024,
      },
    ) as AnyClient;
  }

  private metadata(): grpc.Metadata {
    const md = new grpc.Metadata();
    if (this.apiKey) md.set("authorization", `Bearer ${this.apiKey}`);
    return md;
  }

  /**
   * Per-call deadline. `Infinity` means none, which is what long-lived streams
   * like watches need: grpc-js rejects a far-future deadline rather than
   * reading it as "never".
   */
  private deadline(timeoutMs?: number): grpc.CallOptions {
    const timeout = timeoutMs ?? this.timeoutMs;
    if (!Number.isFinite(timeout)) return {};
    return { deadline: Date.now() + timeout };
  }

  /** Unary call. */
  unary<Req, Res>(
    method: string,
    request: Req,
    timeoutMs?: number,
    signal?: AbortSignal,
  ): Promise<Res> {
    return new Promise((resolvePromise, reject) => {
      let release = () => {};
      const call = this.client[method]!(
        request,
        this.metadata(),
        this.deadline(timeoutMs),
        (err: grpc.ServiceError | null, res: Res) => {
          release();
          if (err) reject(BurrowError.fromGrpc(err));
          else resolvePromise(res);
        },
      );
      release = bindSignal(call, signal);
    });
  }

  /** Server-streaming call, surfaced as an async iterable. */
  async *serverStream<Req, Res>(
    method: string,
    request: Req,
    timeoutMs?: number,
    signal?: AbortSignal,
  ): AsyncGenerator<Res> {
    const call = this.client[method]!(
      request,
      this.metadata(),
      this.deadline(timeoutMs),
    );
    const release = bindSignal(call, signal);
    try {
      yield* iterate<Res>(call);
    } finally {
      release();
    }
  }

  /**
   * Bidirectional call where the client sends `requests` and then closes.
   *
   * Burrow's streaming RPCs all begin with one message that names the sandbox,
   * so this covers exec and upload without exposing a duplex stream.
   */
  async *pipe<Req, Res>(
    method: string,
    requests: Req[],
    timeoutMs?: number,
    signal?: AbortSignal,
  ): AsyncGenerator<Res> {
    const call = this.client[method]!(this.metadata(), this.deadline(timeoutMs));
    const release = bindSignal(call, signal);
    for (const request of requests) call.write(request);
    call.end();
    try {
      yield* iterate<Res>(call);
    } finally {
      release();
    }
  }

  /**
   * Raw bidirectional call, for streams whose input stays open. {@link pipe}
   * closes the request side at once, which an interactive terminal cannot use.
   */
  duplex(method: string): any {
    return this.client[method]!(this.metadata(), {});
  }

  /**
   * Bidirectional call, opened with `first` and left writable: a detached
   * command reads its output as a stream while still able to send a signal.
   */
  openStream<Res>(
    method: string,
    first: unknown,
    timeoutMs?: number,
    signal?: AbortSignal,
  ): { call: any; output: AsyncGenerator<Res> } {
    const call = this.client[method]!(this.metadata(), this.deadline(timeoutMs));
    const release = bindSignal(call, signal);
    call.write(first);
    // Eager, so output produced before the caller reads is buffered.
    const messages = iterate<Res>(call);
    const output = (async function* () {
      try {
        yield* messages;
      } finally {
        release();
      }
    })();
    return { call, output };
  }

  /** Bidi call whose single response is awaited (upload). */
  pipeUnary<Req, Res>(
    method: string,
    requests: Req[],
    timeoutMs?: number,
    signal?: AbortSignal,
  ): Promise<Res> {
    return new Promise((resolvePromise, reject) => {
      let release = () => {};
      const call = this.client[method]!(
        this.metadata(),
        this.deadline(timeoutMs),
        (err: grpc.ServiceError | null, res: Res) => {
          release();
          if (err) reject(BurrowError.fromGrpc(err));
          else resolvePromise(res);
        },
      );
      release = bindSignal(call, signal);
      for (const request of requests) call.write(request);
      call.end();
    });
  }

  /** Claims a share of this connection. Undone by one {@link close}. */
  retain(): this {
    this.refs++;
    return this;
  }

  /** Releases one share; the channel closes when the last one goes. */
  close(): void {
    if (--this.refs <= 0) this.client.close();
  }
}

/**
 * Adapts a gRPC call to an async iterable. Errors arrive on an `error` event
 * rather than rejecting iteration, so they are re-raised into the iterator to
 * keep `for await` and `try/catch` behaving as callers expect.
 */
function iterate<T>(call: any): AsyncGenerator<T> {
  const queue: T[] = [];
  let failure: Error | undefined;
  let done = false;
  let wake: (() => void) | undefined;

  const notify = () => {
    wake?.();
    wake = undefined;
  };

  call.on("data", (msg: T) => {
    queue.push(msg);
    notify();
  });
  call.on("error", (err: grpc.ServiceError) => {
    // CANCELLED after a clean end is how grpc-js reports a stream the caller
    // stopped reading; it is not a failure worth surfacing.
    if (!done || err.code !== grpc.status.CANCELLED) {
      failure = BurrowError.fromGrpc(err);
    }
    done = true;
    notify();
  });
  call.on("end", () => {
    done = true;
    notify();
  });

  // Listeners are attached above rather than on first `next()`: an `error`
  // with nobody listening takes the process down.
  return (async function* () {
    try {
      while (true) {
        while (queue.length > 0) yield queue.shift()!;
        if (failure) throw failure;
        if (done) return;
        await new Promise<void>((r) => {
          wake = r;
        });
      }
    } finally {
      // Abandoning the loop early (break, throw) must not leak the call.
      if (!done) call.cancel();
    }
  })();
}
